Add guide for using `TLSRoute` in operator (#5358)
Kong/developer.konghq.com: app/_how-tos/operator/operator-get-started-gateway-api-4-tlsroute.md

        "op":"add",
        "path":"/spec/template/spec/containers/0/volumeMounts",
        "value":[{
            "mountPath": "/var/run/certs",
            "name": "secret-test",
            "readOnly": true
        }]
    },
    {
        "op":"add",
        "path":"/spec/template/spec/volumes",
        "value":[{
            "name": "secret-test",
            "secret": {
              "defaultMode": 420,
              "secretName": "tls9443.kong.example"
            }
        }]
    }
]'
```

## Route TCP traffic

To reconcile the `TLSRoute`, add a TLS listener to your `Gateway` resource:

```bash
kubectl patch -n kong --type=json gateway kong -p='[
    {
        "op":"add",
        "path":"/spec/listeners/-",
        "value":{
            "name":"stream9443",
            "port":9443,
            "protocol":"TLS",
            "hostname":"tls9443.kong.example",
            "allowedRoutes": {
              "namespaces": {
                "from": "All"
              }
            },
            "tls": {
              "mode": "Passthrough",
              "certificateRefs":[{
                "group":"",
                "kind":"Secret",
                "name":"tls9443.kong.example"

>               }]

Author: Removed `certificateRef` in passthrough listener.

            }
        }
    }
]'

> ```

Author: Changed to `v1`.

Next, create a `TLSRoute`:

```bash
echo "apiVersion: gateway.networking.k8s.io/v1alpha2
kind: TLSRoute
metadata:
  name: echo-tls
  namespace: kong
spec:
  parentRefs:
    - name: kong
      sectionName: stream9443
  hostnames:
    - tls9443.kong.example
  rules:
    - backendRefs:
      - name: echo
        port: 1030
" | kubectl apply -f -

> ```

Author: Already changed to `v1`

---
title: Proxy TLS traffic by SNI using TLSRoute
description: "Use TLSRoute with {{ site.operator_product_name }} to route TCP traffic secured by TLS."

> content_type: how_to

Reviewer: it would be great to have a dedicated content type like `routing` where we are going to group also next pages about GRPC/TCP/UDPRoutes. the final goal is to have a `Routing` section that appears here https://deploy-preview-5358--kongdeveloper.netlify.app/index/operator/#gateway-deployment

Author: Sounds great. I would do it in another PR because this also involves index and content change on the existing pages for `HTTPRoute`.

    {
        "op":"add",
        "path":"/spec/template/spec/containers/0/env/-",
        "value":{
            "name": "TLS_PORT",
            "value": "1030"
        }
    },
    {
        "op":"add",
        "path":"/spec/template/spec/containers/0/env/-",
        "value":{
            "name": "TLS_CERT_FILE",
            "value": "/var/run/certs/tls.crt"
        }
    },
    {
        "op":"add",
        "path":"/spec/template/spec/containers/0/env/-",
        "value":{
            "name": "TLS_KEY_FILE",
            "value": "/var/run/certs/tls.key"
        }
    },
    {
        "op":"add",
        "path":"/spec/template/spec/containers/0/volumeMounts",
        "value":[{
            "mountPath": "/var/run/certs",
            "name": "secret-test",
            "readOnly": true
        }]
    },
    {
        "op":"add",
        "path":"/spec/template/spec/volumes",
        "value":[{
            "name": "secret-test",
            "secret": {
              "defaultMode": 420,
              "secretName": "tls9443.kong.example"
            }
        }]
    }
]'
```

## Route TCP traffic

To reconcile the `TLSRoute`, add a TLS listener to your `Gateway` resource:

```bash

> kubectl patch -n kong --type=json gateway kong -p='[

Reviewer: WDYT about writing the entire Gateway YAML and pass it to a `kubectl apply`? it will be more user readable than a json patch

Author: Done.

The following sections walk through a second `TLSRoute` on a new listener (port `9444`) so it coexists with the Passthrough example above.

### Generate a TLS certificate for the Terminate listener

{% include /k8s/create-certificate.md namespace='kong' hostname='tls9444.kong.example' cert_required=true %}

### Label the certificate Secret

{{ site.operator_product_name }} only watches certificate Secrets that carry the `konghq.com/secret="true"` label. Add the label so Operator picks up the Secret you just created:

```bash
kubectl label secret tls9444.kong.example -n kong konghq.com/secret="true"
```

For more information about how {{ site.operator_product_name }} handles secrets, see the [Secrets reference](/operator/reference/secrets/).

### Add a Terminate TLS listener to the Gateway

Patch the existing `kong` Gateway to append a second TLS listener on port `9444` in `Terminate` mode, referencing the labeled Secret:

```bash

> kubectl patch -n kong --type=json gateway kong -p='[

Reviewer: same here

Author: Done.

---
title: Proxy TLS traffic by SNI using TLSRoute
description: "Use TLSRoute with {{ site.operator_product_name }} to route TCP traffic secured by TLS."
content_type: how_to

permalink: /operator/get-started/gateway-api/create-tlsroute/
series:
  id: operator-get-started-gateway-api
  position: 4

breadcrumbs:
  - /operator/
  - index: operator
    group: Gateway Deployment
  - index: operator
    group: Gateway Deployment
    section: "Get Started"

products:
  - operator

works_on:
  - on-prem
  - konnect

min_version:

>   operator: '2.2.0'

Reviewer 2: ```suggestion
  operator: '2.2'
```

entities: []

tldr:
  q: How do I route TCP traffic with {{ site.operator_product_name }}?
  a: Add a `TLS` listener to your `Gateway` in either `Passthrough` or `Terminate` mode, then create a `TLSRoute` resource. {{ site.operator_product_name }} automatically reconciles the managed `DataPlane` stream listener and proxy Service port, and converts the `TLSRoute` into a {{ site.base_gateway }} [Service](/gateway/entities/service/) and [Route](/gateway/entities/route/).

prereqs:
  skip_product: true

next_steps:
  - text: Learn about Custom resource definitions (CRDs)
    url: /operator/reference/custom-resources/
---

{:.info}

> > **Note**: {{ site.operator_product_name }} only reconciles `TLSRoute` resources using the `gateway.networking.k8s.io/v1` API, which is available in [Gateway API](https://gateway-api.sigs.k8s.io/) 1.5 or later (standard channel). If you installed the CRDs from an earlier release while following [step 1 of this series](/operator/get-started/gateway-api/install/), you should upgrade {{ site.operator_product_name }} to the version that supports `TLSRoute`. It will install or upgrade gateway API CRDs with the appropriate version. If you did not install gateway API CRDs by the {{ site.operator_product_name }} helm release, please install or upgrade them manually:

Reviewer 2: ```suggestion
> {{ site.operator_product_name }} only reconciles `TLSRoute` resources using the `gateway.networking.k8s.io/v1` API, which is available in [Gateway API](https://gateway-api.sigs.k8s.io/) 1.5 or later (standard channel). If you installed the CRDs from an earlier release while following [step 1 of this series](/operator/get-started/gateway-api/install/), upgrade {{ site.operator_product_name }} to the version that supports `TLSRoute`. It will install or upgrade gateway API CRDs with the appropriate version. If you did not install gateway API CRDs by the {{ site.operator_product_name }} helm release, install or upgrade them manually:
```

```shell
kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.5.1/standard-install.yaml --server-side

> ```

Reviewer 2: ```suggestion
> 
> ```shell
> kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.5.1/standard-install.yaml --server-side
> ```
```

## Generate a TLS certificate

{% include /k8s/create-certificate.md namespace='kong' hostname='tls9443.kong.example' cert_required=true %}

## Attach the TLS certificate to the echo Service

If you haven't already deployed the `echo` Service, install it:

```bash
kubectl apply -f {{site.links.web}}/manifests/kic/echo-service.yaml -n kong

> ```

Reviewer 2: ```suggestion
```

Since this is a series, we can assume the user has already done this

> The `echo` Service does not listen on the TLS port by default as it requires a certificate stored in a Kubernetes Secret. Patch the `echo` Deployment to mount the Secret in the pod and set the `TLS_CERT_FILE` and `TLS_KEY_FILE` environment variables to allow the `echo` Service to terminate TLS:

Reviewer 2: ```suggestion
The `echo` Service we deployed in a [previous step](/operator/get-started/gateway-api/create-route/#create-a-backend-service) doesn't listen on the TLS port by default as it requires a certificate stored in a Kubernetes Secret. Patch the `echo` Deployment to mount the Secret in the pod and set the `TLS_CERT_FILE` and `TLS_KEY_FILE` environment variables to allow the `echo` Service to terminate TLS:
```

          from: All
      tls:
        mode: Passthrough' | kubectl apply -f -
```

Next, create a `TLSRoute`:

```bash
echo "apiVersion: gateway.networking.k8s.io/v1
kind: TLSRoute
metadata:
  name: echo-tls
  namespace: kong
spec:
  parentRefs:
    - name: kong
      sectionName: stream9443
  hostnames:
    - tls9443.kong.example
  rules:
    - backendRefs:
      - name: echo
        port: 1030
" | kubectl apply -f -
```

This configuration instructs {{ site.base_gateway }} to forward all traffic it receives on port `9443` to the `echo` Service on port `1030`.

## Validate your configuration

Export the Gateway's external address:

```bash
export PROXY_IP=$(kubectl get gateway kong -n kong -o jsonpath='{.status.addresses[0].value}')
```

You can now access the `echo` Service on port `9443` with SNI `tls9443.kong.example`.

In real-world usage, you would create a DNS record for `tls9443.kong.example` pointing to your proxy Service's public IP address, which causes TLS clients to add an SNI automatically. For this demo, add it manually using the OpenSSL CLI:

```bash
echo "hello" | openssl s_client -connect $PROXY_IP:9443 -servername tls9443.kong.example -quiet 2>/dev/null

> ```

Reviewer 2: When I tested this it took a few tries to get a response, is there a `kubectl wait` command or something similar we can use to make sure the user doesn't get an empty response if the service isn't ready?

The results should look like this:

```text
Welcome, you are connected to node kind-control-plane.
Running on Pod echo-5f44d4c6f9-krnhk.
In namespace default.
With IP address 10.244.0.26.
hello
```

> 

Reviewer 2: ```suggestion
{:.no-copy-code}

```

spec:
  gatewayClassName: kong
  listeners:
    - name: http
      protocol: HTTP
      port: 80
    - name: stream9444
      port: 9444
      protocol: TLS
      hostname: tls9444.kong.example
      allowedRoutes:
        namespaces:
          from: All
      tls:
        mode: Terminate
        certificateRefs:
          - group: ""
            kind: Secret
            name: tls9444.kong.example' | kubectl apply -f -
```

### Create a TLSRoute for the Terminate listener

Because Kong is now terminating TLS, the backend can receive plain TCP. Point the `TLSRoute` at the unmodified `echo` Service on its default plain TCP port `1025`:

```bash
echo "apiVersion: gateway.networking.k8s.io/v1
kind: TLSRoute
metadata:
  name: echo-tls-terminate
  namespace: kong
spec:
  parentRefs:
    - name: kong
      sectionName: stream9444
  hostnames:
    - tls9444.kong.example
  rules:
    - backendRefs:
      - name: echo
        port: 1025
" | kubectl apply -f -
```

### Validate the Terminate listener

Reuse the `$PROXY_IP` you exported earlier and connect to the new listener:

```bash
echo "hello" | openssl s_client -connect $PROXY_IP:9444 -servername tls9444.kong.example -quiet 2>/dev/null

> ```

Reviewer 2: Same question as the previous validation
